shared/crypto/keys.go

```diff
@@ -29,29 +33,59 @@ type Keypair struct {
 }
 
 // SelfSign signs an address to attribute it to a given public key and returns an Identity
-func (k Keypair) SelfSign(suite SigningScheme, address string) (Identity, error) {
-	message := append(k.Public, []byte(address)...)
+func (k Keypair) SelfSign(suite SigningScheme, address string, nonce uint32) (Identity, error) {
+	message, err := messageFor(k.Public, address, nonce)
+	if err != nil {
+		return Identity{}, err
+	}
 
 	signature, err := suite.Sign(k, message)
 	if err != nil {
 		return Identity{}, err
 	}
 
 	return Identity{
-		Address:   address,
-		Public:    k.Public,
-		Signature: signature,
+		Address:        address,
+		Public:         k.Public,
+		Signature:      signature,
+		ValidatorNonce: nonce,
 	}, nil
 }
 
 type Identity struct {
-	Address   string `json:"address"`
-	Public    []byte `json:"public"`
-	Signature []byte `json:"signature"`
+	ValidatorNonce uint32 `json:"validator_nonce"`
+	Address        string `json:"address"`
+	Public         []byte `json:"public"`
+	Signature      []byte `json:"signature"`
 }
 
 // Verify checks the signature for a given identity is valid, if e.g. pulled from a remote file
 func (i Identity) Verify(suite SigningScheme) error {
-	m := append(i.Public, []byte(i.Address)...)
+	m, err := messageFor(i.Public, i.Address, i.ValidatorNonce)
+	if err != nil {
+		return err
+	}
 	return suite.Verify(m, i.Public, i.Signature)
 }
+
+func messageFor(publicKey []byte, address string, validatorNonce uint32) ([]byte, error) {
+	buf := new(bytes.Buffer)
+
+	if err := binary.Write(buf, binary.BigEndian, publicKey); err != nil {
+		return nil, err
+	}
+
+	if err := binary.Write(buf, binary.BigEndian, []byte(address)); err != nil {
+		return nil, err
+	}
+
+	if err := binary.Write(buf, binary.BigEndian, validatorNonce); err != nil {
+		return nil, err
+	}
+
+	out := sha256.New()
+	if _, err := out.Write(buf.Bytes()); err != nil {
+		return nil, err
+	}
+	return out.Sum(nil), nil
```

Comment:
This looks more like a "digestFor" or "hashFrom" function.
Also, it might be a good idea to have some sort of domain separation tags to avoid risks of clashes, or is this formatting dictated by the SSV side of things?

Reply:
That being said, I'd be hard pressed to find a collision, but since these are user-provided values...